### PR TITLE
Regular expression (regexp) matcher

### PR DESCRIPTION
**Pull request by vnxme**
### Summary

The idea behind `regexp` matcher is to allow low-level packet matching. It's pretty simple under the hood, but actually very powerful. With this matcher Caddy can multiplex virtually any protocol, if a regular expression syntax is enough to distinguish the desired one from others.

### Code examples

Proxy a connection starting with a 0xFFEEDD000000 sequence to another machine
```json
{
  "apps": {
    "layer4": {
      "servers": {
        "example": {
          "listen": ["0.0.0.0:443"],
          "routes": [
            {
              "handle": [
                {
                  "handler": "proxy",
                  "upstreams": [
                    {"dial": ["another.machine.local:456"]}
                  ]
                }
              ],
              "match": [
                {"regexp": {"count": 6, "pattern": "^\\xFF\\xEE\\xDD\\x00\\x00\\x00$"}}
              ]
            },
            {
              "handle": [
                {
                  "handler": "tls"
                },
                {
                  "handler": "proxy",
                  "upstreams": [
                    {"dial": ["localhost:8080"]}
                  ]
                }
              ],
              "match": [
                {"tls": {}}
              ]
            }
          ]
        }
      }
    }
  }
}
```

### CODE REVIEW

File: README.md
Comment (mholt) [inline README.md:41]:
```suggestion
- **layer4.matchers.regexp** - matches connections that have the first packet bytes matching a regular expression.
```

File: modules/l4regexp/matcher.go
Comment (mholt) [inline modules/l4regexp/matcher.go:1]: Typically, throughout the repo, we use the same copyright (a "project" copyright as it were) and leave it to the commit history to assign credit to individual contributors.

That said, I'm not a lawyer nor do I have strong feelings about this in this case in this repo, but I could see if we end up moving this code into Caddy mainline it could be confusing with multiple copyright assignments throughout the code.

I definitely want to credit you though so what if we put a comment below the license statement and say "This file contributed by @vnxme" or something?

**Review by mholt — COMMENTED** [review]
This looks great! I'm excited to see how people use it.

File: modules/l4regexp/matcher.go
Comment (vnxme) [reply modules/l4regexp/matcher.go:1]: Well, I just copied your typical license block and put my name instead. I think it only indicates that the file is written by me, but it doesn’t restrict any use in the project. Neither was it my intention to put any restrictions.

For example, please see how opnsense copyrights look like [here](https://github.com/opnsense/core/blob/efea5429524e847b0c1371da6c4ff666f54c2459/src/etc/rc#L3). If more than one person contributed to the file, they put multiple names of contributors sorted by years of changes.

Comment (mholt) [reply modules/l4regexp/matcher.go:1]: Oh interesting, ok. That's probably fine then.

**Review by mholt — APPROVED** [review]
LGTM! Anything else before we merge it?

**Comment by vnxme** [comment]
Rebased onto master and added integration tests. I think, it's ready to be merged now.

**Comment by mholt** [comment]
Awesome! Thanks!

**Pull request merged**
